### DISCUSSION TITLE
Trying to generate a static site using Nuxt3 + Strapi CMS with fully static content for production and useFetch API content refreshes in dev mode

### DISCUSSION THREAD

**Discussion by noahps** [nuxt/nuxt#23236] [Questions] [answered]
My first nuxt project and I'm trying to setup a site that can be deployed completely pre-rendered for production, but using live content in dev mode. The reason for this is that the content is being put together by content creators and I'd like them to be able to see their edits during development. For my setup, I have:

- a strapi server is running on a temporary EC2 machine that may be turned off once development is totally signed off and content no longer needs to change.
- local dev environment where I am building/testing using nuxt dev with `ssr: false`
- dev staging environment that hosts static files that are rendered by using `nuxt generate` with `ssr: false`. I would like the pages to send requests to the strapi api to show the latest content without needing to re-render the site
- production hosting environment that hosts static files that are rendered by using `nuxt generate` with `ssr: true`. For production deployment I'm trying to get the site to render all pages with `ssr: true` so that no further calls are made to the stapi api.

I am able to get things to work either one way or the other, but not both. My first thought was that using `useFetch` would allow me to render the content for both cases, and it seems fine on initial page load. But when navigating through the site, subsequent page loads cause useFetch to refresh the content from the API rather than using the already cached content. This would seem ok, except as noted the plan is to turn off the strapi server once the site and content are stable.

To solve this problem, I'm trying a combination of `useNuxtData` and `useFetch`.  Here's a code snippet:

**index.vue**
```
// api content
const contentType = "contact";
const key = `${contentType}_${locale.value}`;

const { data: pageContent } = useNuxtData(key);

if (!process.client || useRuntimeConfig().public.ssr === false) {
  // only fetch data in dev environment or during generate
  const apiURL = getStrapiLink(contentType);

  const { data } = await useFetch(apiURL, {
    key: key,
    default: () => {
      // load the cached value by default to prevent api refreshes
      return pageContent.value;
    },
  });
}
```

**nuxt.config.ts**
```
    ...
    runtimeConfig: {
        public: {
            ...
            ssr: process.env.APP_USE_SSR == 'TRUE' ? true : false,
        },
    },
    ssr: process.env.APP_USE_SSR == 'TRUE' ? true : false,
    ...
```

This is giving me the desired result for a final production environment when I have `ssr: true`. However, for my dev build with `ssr: false` I get the error:

_Uncaught (in promise) TypeError: Cannot read properties of null (reading 'data')_

Which is being triggered by part of the template trying to render an element like this:
```
<p>{{ pageContent.data.attributes.ContactPageIntro }}</p>
```

The content returned by the strapi api comes in the shape:
```
{
  "data": {
    "id": 1,
    "attributes": {
      "ContactPageIntro": "Please contact us ... "
    }
  },
  "meta": {
    
  }
}
```

The fetch is firing and the data is coming, but not before the page tries to render so it's throwing 500 errors.

Any advice or suggestions how to make this work would be much appreciated!

**Comment by yassilah** [comment, marked as answer]
Correct me if I'm wrong as I haven't tested it myself but I think you might be looking for something along the lines of this:

**index.vue**
```ts
const contentType = "contact";
const key = `${contentType}_${locale.value}`;
const apiURL = getStrapiLink(contentType);

const { data } = await useFetch(apiURL, {
    key: key,
    watch: false
};
```

**nuxt.config.ts**
```ts
export default defineNuxtConfig({
    ssr: false,
    experimental: {
        payloadExtraction: true
    },
    $development: {
        experimental: {
            payloadExtraction: false
        }
    },
    $production: {
        nitro: {
            static: true
        }
    }
})
```

**Command**
```bash
npm run build
```

**Development**: no SSR + no payload extraction
**Staging**: no SSR + payload extraction
**Production**: full-static

**Reply by noahps** [reply]
This is great! Thanks @yassilah . I made a slight modification because `nuxt build` and `nuxt generate` commands force the NODE_ENV to production. Perhaps there's a proper way to do that, but what I'm doing instead is using .dotenv commands to load my own .env files for the staging environment I need.  I have my component working the way you suggested, and my nuxt.config.ts looks like this:

```ts
export default defineNuxtConfig({
    // ...
    experimental: {
        // conditionally set payloadExtraction depending on staging environment
        payloadExtraction: process.env.APP_USE_SSR == 'TRUE' ? true : false,
    },
    nitro: {
        // conditionally set nitro static depending on staging environment
        static: process.env.APP_USE_SSR == 'TRUE' ? true : false,
      },
    // conditionally set ssr setting depending on staging environment
    ssr: process.env.APP_USE_SSR == 'TRUE' ? true : false
    // ...
})
```

Thanks so much for your help!

**Reply by yassilah** [reply]
Glad I could help 😊
